sound/soc/sof/ipc4-pcm.c

```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
 	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;
 	struct ipc4_pipeline_set_state_data *data;
+	struct snd_soc_dapm_widget_list *list;
 	struct snd_sof_widget *pipe_widget;
+	struct snd_soc_dapm_widget *widget;
 	struct sof_ipc4_pipeline *pipeline;
+	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
+	int num_widgets;
 	int ret;
 	int i, j;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
+        list = spcm->stream[substream->stream].list;
+        for_each_dapm_widgets(list, num_widgets, widget) {
+		swidget = widget->dobj.private;
+
+		if (!swidget)
+                        continue;
+		/* find pipeline widget for the pipeline that this widget belongs to */
+		pipe_widget = swidget->pipe_widget;
+		pipeline = (struct sof_ipc4_pipeline *)pipe_widget->private;
+
+		if (pipeline->use_chain_dma) {
+			struct sof_ipc4_msg *msg = &pipeline->msg;
+			/* only handle the AIF widget for chain DMA */
+			if (!WIDGET_IS_AIF(swidget->id))
+				continue;
+
+			if (state == SOF_IPC4_PIPE_RUNNING) {
+				/* Set allocate and enable bits to configure chain DMA. The
+				 * other msg bits are set in sof_ipc4_prepare_copier_module()
+				 * and in sof_ipc4_dai_config(), and cleared in
+				 * sof_ipc4_unprepare_copier_module().
+				 */
+				msg->primary &= ~(SOF_IPC4_MSG_TYPE_MASK | SOF_IPC4_MSG_DIR_MASK |
+						  SOF_IPC4_MSG_TARGET_MASK);
+				msg->primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_CHAIN_DMA);
+				msg->primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+				msg->primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+				msg->primary |= SOF_IPC4_GLB_CHAIN_DMA_ALLOCATE_MASK;
+				msg->primary |= SOF_IPC4_GLB_CHAIN_DMA_ENABLE_MASK;
```

Comment:
probably need comments to explain why we have to do allocate and enabled at the same time